File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/v4/ApiSearchService.java
/*
 * Copyright © 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.rest.api.service.v4;

import io.gravitee.common.data.domain.Page;
import io.gravitee.definition.model.DefinitionVersion;
import io.gravitee.repository.management.model.Api;
import io.gravitee.rest.api.model.api.ApiQuery;
import io.gravitee.rest.api.model.common.Pageable;
import io.gravitee.rest.api.model.common.Sortable;
import io.gravitee.rest.api.model.v4.api.ApiEntity;
import io.gravitee.rest.api.model.v4.api.GenericApiEntity;
import io.gravitee.rest.api.service.common.ExecutionContext;
import io.gravitee.rest.api.service.search.query.QueryBuilder;
import java.util.Collection;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

/**
 * @author Guillaume LAMIRAND (guillaume.lamirand at graviteesource.com)
 * @author GraviteeSource Team
 */
public interface ApiSearchService {
    ApiEntity findById(ExecutionContext executionContext, String apiId);

    GenericApiEntity findGenericById(final ExecutionContext executionContext, final String apiId);

    Set<GenericApiEntity> findAllGenericByEnvironment(ExecutionContext executionContext);

    Api findV4RepositoryApiById(ExecutionContext executionContext, String apiId);

    Api findRepositoryApiById(ExecutionContext executionContext, String apiId);

    Optional<String> findIdByEnvironmentIdAndCrossId(final String environment, final String crossId);

    boolean exists(final String apiId);

    Set<GenericApiEntity> findGenericByEnvironmentAndIdIn(ExecutionContext executionContext, Set<String> apiIds);

    Collection<GenericApiEntity> search(final ExecutionContext executionContext, final ApiQuery query);

    Collection<GenericApiEntity> search(ExecutionContext executionContext, ApiQuery query, boolean excludeDefinitionV4);

    Page<GenericApiEntity> search(
        final ExecutionContext executionContext,
        final String userId,
        final boolean isAdmin,
        final QueryBuilder<ApiEntity> queryBuilder,
        final Pageable pageable,
        final boolean mapToFullGenericApiEntity,
        final boolean manageOnly
    );

    Collection<String> searchIds(
        final ExecutionContext executionContext,
        final String query,
        Map<String, Object> filters,
        final Sortable sortable
    );

    Collection<String> searchIds(
        ExecutionContext executionContext,
        String query,
        Map<String, Object> filters,
        Sortable sortable,
        Collection<DefinitionVersion> excludeDefinitionVersions
    );
}

File: gravitee-apim-portal-webui-next/projects/gravitee-markdown/src/lib/services/api-search.service.ts
/*
 * Copyright (C) 2024 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { HttpClient } from '@angular/common/http';
import { Injectable, Inject } from '@angular/core';
import { Observable } from 'rxjs';

import { GRAVITEE_MARKDOWN_BASE_URL } from './configuration';

export interface ApiSearchResponse {
  data: Api[];
  metadata: {
    total: number;
    page: number;
    size: number;
  };
}

export interface Api {
  id: string;
  name: string;
  description?: string;
  version: string;
  state: string;
  contextPath: string;
  tags?: string[];
  categories?: string[];
}

@Injectable({
  providedIn: 'root',
})
export class ApiSearchService {
  constructor(
    private readonly http: HttpClient,
    @Inject(GRAVITEE_MARKDOWN_BASE_URL) private baseURL: string,
  ) {}

  search(page = 1, category: string = 'all', q: string = '', size = 9): Observable<ApiSearchResponse> {
    return this.http.post<ApiSearchResponse>(`${this.baseURL}/apis/_search`, null, {
      params: {
        page: page.toString(),
        category: category !== 'all' && category !== undefined ? category : '',
        size: size.toString(),
        q: q,
      },
    });
  }

  details(apiId: string): Observable<Api> {
    return this.http.get<Api>(`${this.baseURL}/apis/${apiId}`);
  }
} 